### DISCUSSION TITLE
How to validate array length and each item with custom validation?

### DISCUSSION THREAD

**Discussion by tparvi** [react-hook-form/react-hook-form#7815] [Q&A] [answered]
I am using custom validation (so just function assigned to form's resolver prop) to validate data. How can I validate both the length of an array and individual items? E.g. I have an array of persons and I would like to validate two things and display the validation errors to the user:
- User has to have at least two persons in the array
- Each person must have a name

If I want to validate individual persons my model is as follows:

```
type Person = {
    name: string;
}

type FormModel = {
    persons: Person[]
}

```
In VS Code I can see that I can validate each person because in the validation result the persons is an array of FieldErrors. This works correctly. But how do I return error that the length of the array is not correct? There is no place for it.

![image](https://user-images.githubusercontent.com/871053/153822345-c36e7864-1141-4584-a219-a3be348b7855.png)

If I use `NestedValue` I can only return single FieldError for the whole array. So I can validate the length of an array but I cannot validate individual items.

```
type FormModel = {
    persons: NestedValue<Person[]>
}
```

![image](https://user-images.githubusercontent.com/871053/153822388-00343e5b-bca9-4ff3-8e81-5063c5c74528.png)

**Comment by Moshyfawn** [comment, marked as answer]
You can definetly add whatever validation to your array and it's items using resolvers
Example with [Zod](https://github.com/colinhacks/zod)
```js
const itemSchema = z.object({
  id: string().uuid(),
  name: string().optional()
})

const formSchema = `.object({
  array: z.array(itemSchema).min(5, "Please, add at least 5 items")
})
```
When it comes to displaying the array error `errors.array?.message`, the field is there, but ATM RHF has the TS limitation to correctly assert it's path. So, the only way is to `@ts-expect-error`

**Reply by tparvi** [reply]
We are not using zod because there is lot of validation where validation of field1 depends on value of other fields. So we are just using function that returns `ResolverResult<T>`. The issue is also when creating these validation messages as you can only return messages for specific items inside the array or one message for the whole array:

```
type Value = {
  name: string;
}

type FormModel = {
  value1: Value[]
  value2: NestedValue<Value[]>
}

function validate(input:FormModel):ResolverResult<FormModel> {
  const result:FieldErrors<FormModel> = {}

  // Since value1 is an array I can only create validation messages that are for specific item
  // in the array. I Cannot return error about min length of array.
  result.value1 = [
    {
      name: {
        type: 'validate',
        message: 'Value is invalid'
      }
    }
  ]

  // Since value2 is NestedValue I can only validate e.g. array length. Cannot return errors for specific array items.
  result.value2 = {
    type: 'validate',
    message: 'Sorry you have to select three values'
  }

  return {
    errors: result,
    values: input
  }
}
```

**Reply by bluebill1049** [reply]
TS: This is a design limitation, we simply can't support both object and array in a single errors node. I have some idea about this but nothing is concrete or solid.

**Reply by Moshyfawn** [reply]
> We are not using zod because there is lot of validation where validation of field1 depends on value of other fields

As a side note, you can create dependent field validation with [Zod](https://github.com/colinhacks/zod) by utilizing [the `.refine()` method](https://github.com/colinhacks/zod#refine)

**Reply by tparvi** [reply]
Thanks for the quick reply! It is not a huge issue. In case someone else runs into this I am solving it by adding additional `onlyUsedForArrayValidation` field and using it to store the min/max length validation message. There might be a cleaner way to achieve the same result but at least the intent comes across from the following code:

```
type Value = {
  name: string;
}

type FormModel = {
  onlyUsedForArrayValidation: string;
  value: Value[];
}

function validate(input:FormModel):ResolverResult<FormModel> {
  const result:FieldErrors<FormModel> = {}

  // Validate individual items and store the validation messages like this
  result.value = [
    {
      name: {
        type: 'validate',
        message: 'Value is invalid'
      }
    }
  ]

  // If the array itself (min/max length is invalid), store error like this
  result.onlyUsedForArrayValidation= {
    type: 'validate',
    message: 'Sorry you have to select three values'

  }

  // return result....
}
```
